database/gorm/query.go

```diff
@@ -643,122 +703,352 @@ func (r *QueryImpl) Update(column any, value ...any) (*ormcontract.Result, error
 }
 
 func (r *QueryImpl) UpdateOrCreate(dest any, attributes any, values any) error {
-	if err := r.refreshConnection(dest); err != nil {
+	query, err := r.refreshConnection(dest)
+	if err != nil {
 		return err
 	}
-	res := r.instance.Assign(values).FirstOrInit(dest, attributes)
+
+	query = query.buildConditions()
+
+	res := query.instance.Assign(values).FirstOrInit(dest, attributes)
 	if res.Error != nil {
 		return res.Error
 	}
 	if res.RowsAffected > 0 {
-		return r.Save(dest)
+		return query.Save(dest)
 	}
 
-	return r.Create(dest)
+	return query.Create(dest)
 }
 
 func (r *QueryImpl) Where(query any, args ...any) ormcontract.Query {
-	tx := r.instance.Where(query, args...)
+	conditions := r.conditions
+	conditions.where = append(r.conditions.where, Where{
+		query: query,
+		args:  args,
+	})
 
-	return NewQueryImplByInstance(tx, r)
+	return r.setConditions(conditions)
 }
 
-func (r *QueryImpl) WithoutEvents() ormcontract.Query {
-	return NewQueryImplByInstance(r.instance, &QueryImpl{
-		config:        r.config,
-		withoutEvents: true,
+func (r *QueryImpl) With(query string, args ...any) ormcontract.Query {
+	conditions := r.conditions
+	conditions.with = append(r.conditions.with, With{
+		query: query,
+		args:  args,
 	})
+
+	return r.setConditions(conditions)
+}
+
+func (r *QueryImpl) WithoutEvents() ormcontract.Query {
+	conditions := r.conditions
+	conditions.withoutEvents = true
+
+	return r.setConditions(conditions)
 }
 
 func (r *QueryImpl) WithTrashed() ormcontract.Query {
-	tx := r.instance.Unscoped()
+	conditions := r.conditions
+	conditions.withTrashed = true
 
-	return NewQueryImplByInstance(tx, r)
+	return r.setConditions(conditions)
 }
 
-func (r *QueryImpl) With(query string, args ...any) ormcontract.Query {
-	if len(args) == 1 {
-		switch arg := args[0].(type) {
-		case func(ormcontract.Query) ormcontract.Query:
-			newArgs := []any{
-				func(tx *gormio.DB) *gormio.DB {
-					query := arg(NewQueryImplByInstance(tx, r))
-
-					return query.(*QueryImpl).instance
-				},
-			}
+func (r *QueryImpl) buildConditions() *QueryImpl {
```

Comment:
The original Gorm instance of QueryImpl will not be changed, to avoid the conditions chaotic.